tests/kernel/zync/src/main.c

```diff
@@ -472,7 +474,52 @@ ZTEST(zync_tests_1cpu, test_abort_recover)
 	k_sched_unlock();
 
 	k_sleep(K_TICKS(1));
-	zassert_equal(awoken_count, 1, "didn't wake up");
+	zassert_equal(awoken_count, 1, "replacement thread didn't wake up");
+}
+
+static void timeout_wakeup(void *pa, void *pb, void *pc)
+{
+	int32_t ticks = k_ms_to_ticks_ceil32(300);
+	k_timeout_t timeout = K_TICKS(ticks);
+
+	int64_t start = k_uptime_ticks();
+	int32_t ret = k_sem_take(&wrapped_sem, timeout);
+	int64_t end = k_uptime_ticks();
+
+	zassert_equal(ret, -EAGAIN, "k_sem_take() should return -EAGAIN");
+
+	int64_t dt = end - start;
+
+	if (IS_ENABLED(CONFIG_ZYNC_STRICT_TIMEOUTS)) {
+		zassert_true(dt >= ticks, "didn't wait long enough: dt == %d", dt);
+	} else {
+		/* 3-tick threshold for 2 context switches and a 1
+		 * tick sleep in the main thread.
+		 */
+		zassert_true(dt <= 3, "should have woken up immediately");
+	}
+}
+
+/* Tests the zync pair retry behavior */
+ZTEST(zync_tests_1cpu, test_early_wakeup)
```

Comment:
I tried a different test on a single core system and got something that seemed counter-intuitive.

0. Setup: Two threads waiting on a semaphore (Tmed and Tlow)
1. Thigh gives the semaphore. This should wake Tmed, but not schedule it.
2. Thigh re-takes semaphore. This should succeed.
3. Thigh is busy for a while.
4. Thigh gives semaphore AFTER Tmed was supposed to originally time out. (I expect this to wake Tlow)
5. Thigh sleeps for a while so that Tmed and Tlow can run. Tmed is scheduled in first.
6. Expectation: Tmed runs but its k_sem_take() fails. Tmed finishes and lets Tlow run.
7. Expectation: Tlow finishes and its k_sem_take() succeeds.

I am observing the following ...
Tmed gets the semaphore. Tlow does not.